pipeline/authn/authenticator_oauth2_client_credentials.go

```diff
@@ -123,8 +123,23 @@ func (a *AuthenticatorOAuth2ClientCredentials) Authenticate(r *http.Request, ses
 		oauth2.HTTPClient,
 		c.Client,
 	))
+
 	if err != nil {
-		return errors.Wrapf(helper.ErrUnauthorized, err.Error())
+		if rErr, ok := err.(*oauth2.RetrieveError); ok {
+			if rErr.Response.StatusCode == http.StatusServiceUnavailable {
+				return errors.Wrapf(helper.ErrUpstreamServiceNotAvailable, err.Error())
+			} else if rErr.Response.StatusCode == http.StatusInternalServerError {
+				return errors.Wrapf(helper.ErrUpstreamServiceInternalServerError, err.Error())
+			} else if rErr.Response.StatusCode == http.StatusGatewayTimeout {
+				return errors.Wrapf(helper.ErrUpstreamServiceTimeout, err.Error())
+			} else if rErr.Response.StatusCode == http.StatusNotFound {
+				return errors.Wrapf(helper.ErrUpstreamServiceNotFound, err.Error())
+			} else if rErr.Response.StatusCode == http.StatusUnauthorized {
+				return errors.Wrapf(helper.ErrUnauthorized, err.Error())
+			}
```

Comment:
A default case is missing. If that case were to happen, this would let the request pass even if an error occurred. I recommend using switch with a defaults!